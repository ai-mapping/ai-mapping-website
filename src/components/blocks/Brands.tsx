'use client';

import React, { useState } from 'react';
import Image from 'next/image';
import Link from 'next/link';
import Section from '@/components/layout/Section';

// Define types for brand items
export type BrandItem = {
  name: string;
  isLarge?: boolean;
  logoSrc: string; // Changed to string for URL-based images
  logoAlt?: string; // Alt text for accessibility
  logoHeight: number;
  logoWidth?: number; // Optional, Next.js can calculate from aspect ratio
  href: string; // Link to catalog section
};

// Main props interface
export type BrandsProps = {
  heading: string;
  largeBrands: BrandItem[];
  smallBrands: BrandItem[];
};

// Default props
export const defaultBrandsProps: BrandsProps = {
  heading: 'Compatible with Leading Frontend Frameworks',
  largeBrands: [
    {
      name: 'React',
      isLarge: true,
      logoSrc: '/images/placeholders/react-logo.svg',
      logoAlt: 'React logo',
      logoHeight: 70,
      href: '/frameworks/react'
    },
    {
      name: 'Next.js',
      isLarge: true,
      logoSrc: '/images/placeholders/nextjs-logo.png',
      logoAlt: 'Next.js logo',
      logoHeight: 80,
      href: '/frameworks/nextjs'
    }
  ],
  smallBrands: [
    {
      name: 'Tailwind CSS',
      logoSrc: '/images/placeholders/tailwind-logo.svg',
      logoAlt: 'Tailwind CSS logo',
      logoHeight: 40,
      href: '/tools/tailwind'
    },
    {
      name: 'TypeScript',
      logoSrc: '/images/placeholders/typescript-logo.svg',
      logoAlt: 'TypeScript logo',
      logoHeight: 40,
      href: '/tools/typescript'
    },
    {
      name: 'Gatsby',
      logoSrc: '/images/placeholders/gatsby-logo.svg',
      logoAlt: 'Gatsby logo',
      logoHeight: 40,
      href: '/frameworks/gatsby'
    },
    {
      name: 'Remix',
      logoSrc: '/images/placeholders/remix-logo.svg',
      logoAlt: 'Remix logo',
      logoHeight: 40,
      href: '/frameworks/remix'
    },
    {
      name: 'Astro',
      logoSrc: '/images/placeholders/astro-logo.svg',
      logoAlt: 'Astro logo',
      logoHeight: 40,
      href: '/frameworks/astro'
    }
  ]
};

const Brands = ({
  heading = defaultBrandsProps.heading,
  largeBrands = defaultBrandsProps.largeBrands,
  smallBrands = defaultBrandsProps.smallBrands
}: Partial<BrandsProps>) => {
  // State to track image loading errors
  const [imageErrors, setImageErrors] = useState<Record<string, boolean>>({});

  // Handle image error
  const handleImageError = (brandName: string) => {
    console.log(`Failed to load logo for: ${brandName}`);
    setImageErrors(prev => ({
      ...prev,
      [brandName]: true
    }));
  };

  return (
    <Section background="surface" paddingY="md" containerWidth="container">
        <h2 className="text-2xl md:text-3xl font-semibold text-center mb-10 text-white">
          {heading}
        </h2>

        <div className="flex flex-wrap justify-center items-center gap-8 mb-8">
          {largeBrands.map((brand, index) => (
            <div key={`large-brand-${index}`} className="w-1/3 md:w-auto">
              <Link
                href={brand.href}
                className="block transition-all duration-300 hover:scale-105 hover:shadow-lg"
                title={`View details for ${brand.name}`}
              >
                <div className="bg-neutral-light p-2 rounded-lg flex items-center justify-center h-20 md:h-24 md:w-48 border border-neutral-dark shadow-md">
                  {!imageErrors[brand.name] ? (
                    <Image
                      src={brand.logoSrc}
                      alt={brand.logoAlt || `${brand.name} logo`}
                      height={brand.logoHeight}
                      width={brand.logoWidth || 120}
                      className="h-auto max-h-full w-auto"
                      priority={index < 2} // Prioritize loading for first two logos
                      quality={90}
                      onError={() => handleImageError(brand.name)}
                    />
                  ) : (
                    <div className="text-center font-medium text-neutral">
                      {brand.name}
                    </div>
                  )}
                </div>
              </Link>
            </div>
          ))}
        </div>

        <div className="flex flex-wrap justify-center items-center gap-6 md:gap-8 my-6">
          {smallBrands.map((brand, index) => (
            <div key={`small-brand-${index}`} className="w-1/4 md:w-auto">
              <Link
                href={brand.href}
                className="block transition-all duration-300 hover:scale-105 hover:shadow-md"
                title={`View details for ${brand.name}`}
              >
                <div className="bg-neutral-light p-3 rounded-lg flex items-center justify-center h-16 md:h-16 md:w-32 border border-neutral-dark shadow-sm">
                  {!imageErrors[brand.name] ? (
                    <Image
                      src={brand.logoSrc}
                      alt={brand.logoAlt || `${brand.name} logo`}
                      height={brand.logoHeight}
                      width={brand.logoWidth || 80}
                      className="h-auto max-h-full w-auto"
                      quality={85}
                      onError={() => handleImageError(brand.name)}
                    />
                  ) : (
                    <div className="text-center font-medium text-neutral">
                      {brand.name}
                    </div>
                  )}
                </div>
              </Link>
            </div>
          ))}
        </div>
    </Section>
  );
};

export default Brands;