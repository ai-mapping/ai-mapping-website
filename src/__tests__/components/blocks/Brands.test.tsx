import React from 'react';
import { render, screen } from '@testing-library/react';
import Brands, { defaultBrandsProps } from '@/components/blocks/Brands';

// Mock images
jest.mock('@/components/ui/logos/caterpillar-logo.png', () => ({
  src: '/caterpillar-logo.png',
  height: 70,
  width: 140
}));

jest.mock('@/components/ui/logos/komatsu-logo.png', () => ({
  src: '/komatsu-logo.png',
  height: 80,
  width: 160
}));

jest.mock('@/components/ui/logos/zf-logo.png', () => ({
  src: '/zf-logo.png',
  height: 40,
  width: 80
}));

jest.mock('@/components/ui/logos/BOMAG-logo.png', () => ({
  src: '/BOMAG-logo.png',
  height: 40,
  width: 80
}));

jest.mock('@/components/ui/logos/Mobil.png', () => ({
  src: '/Mobil.png',
  height: 40,
  width: 80
}));

jest.mock('@/components/ui/logos/hyva-logo.png', () => ({
  src: '/hyva-logo.png',
  height: 40,
  width: 80
}));

jest.mock('@/components/ui/logos/FAG-logo.png', () => ({
  src: '/FAG-logo.png',
  height: 40,
  width: 80
}));

describe('Brands Component', () => {
  it('renders with default props correctly', () => {
    const { container } = render(<Brands />);
    
    // Check if heading is rendered
    expect(screen.getByText(defaultBrandsProps.heading)).toBeInTheDocument();
    
    // Check if all large brands are rendered
    defaultBrandsProps.largeBrands.forEach((brand) => {
      expect(screen.getByTitle(`View details for ${brand.name}`)).toBeInTheDocument();
      expect(screen.getByTestId(`next-image-${brand.logoAlt || `${brand.name} logo`}`)).toBeInTheDocument();
    });
    
    // Check if all small brands are rendered
    defaultBrandsProps.smallBrands.forEach((brand) => {
      expect(screen.getByTitle(`View details for ${brand.name}`)).toBeInTheDocument();
      expect(screen.getByTestId(`next-image-${brand.logoAlt || `${brand.name} logo`}`)).toBeInTheDocument();
    });
    
    // Snapshot test
    expect(container).toMatchSnapshot();
  });
  
  it('renders with custom props correctly', () => {
    const customProps = {
      heading: 'Custom Brands Heading',
      largeBrands: [
        {
          name: 'Custom Brand',
          logoSrc: { src: '/custom-logo.png', height: 60, width: 120 },
          logoAlt: 'Custom brand logo',
          logoHeight: 60,
          href: '/custom-link'
        }
      ],
      smallBrands: [
        {
          name: 'Small Custom Brand',
          logoSrc: { src: '/small-custom-logo.png', height: 30, width: 60 },
          logoAlt: 'Small custom brand logo',
          logoHeight: 30,
          href: '/small-custom-link'
        }
      ]
    };
    
    const { container } = render(<Brands {...customProps} />);
    
    // Check if custom heading is rendered
    expect(screen.getByText(customProps.heading)).toBeInTheDocument();
    
    // Check if custom large brand is rendered
    expect(screen.getByTitle(`View details for ${customProps.largeBrands[0].name}`)).toBeInTheDocument();
    expect(screen.getByTestId(`next-image-${customProps.largeBrands[0].logoAlt}`)).toBeInTheDocument();
    
    // Check if custom small brand is rendered
    expect(screen.getByTitle(`View details for ${customProps.smallBrands[0].name}`)).toBeInTheDocument();
    expect(screen.getByTestId(`next-image-${customProps.smallBrands[0].logoAlt}`)).toBeInTheDocument();
    
    // Snapshot test
    expect(container).toMatchSnapshot();
  });
});