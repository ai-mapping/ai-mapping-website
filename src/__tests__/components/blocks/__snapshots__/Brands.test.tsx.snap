// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Brands Component renders with custom props correctly 1`] = `
<div>
  <section
    class="bg-surface py-10"
  >
    <div
      class="container mx-auto px-4 md:px-6 lg:px-8"
    >
      <h2
        class="text-2xl md:text-3xl font-semibold text-center mb-10 text-white"
      >
        Custom Brands Heading
      </h2>
      <div
        class="flex flex-wrap justify-center items-center gap-8 mb-8"
      >
        <div
          class="w-1/3 md:w-auto"
        >
          <a
            class="block transition-all duration-300 hover:scale-105 hover:shadow-lg"
            data-testid="next-link"
            href="/custom-link"
            title="View details for Custom Brand"
          >
            <div
              class="bg-neutral-light p-2 rounded-lg flex items-center justify-center h-20 md:h-24 md:w-48 border border-neutral-dark shadow-md"
            >
              <img
                alt="Custom brand logo"
                class="h-auto max-h-full w-auto"
                data-testid="next-image-Custom brand logo"
                height="60"
                src="/custom-logo.png"
                width="120"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        class="flex flex-wrap justify-center items-center gap-6 md:gap-8 my-6"
      >
        <div
          class="w-1/4 md:w-auto"
        >
          <a
            class="block transition-all duration-300 hover:scale-105 hover:shadow-md"
            data-testid="next-link"
            href="/small-custom-link"
            title="View details for Small Custom Brand"
          >
            <div
              class="bg-neutral-light p-3 rounded-lg flex items-center justify-center h-16 md:h-16 md:w-32 border border-neutral-dark shadow-sm"
            >
              <img
                alt="Small custom brand logo"
                class="h-auto max-h-full w-auto"
                data-testid="next-image-Small custom brand logo"
                height="30"
                src="/small-custom-logo.png"
                width="80"
              />
            </div>
          </a>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`Brands Component renders with default props correctly 1`] = `
<div>
  <section
    class="bg-surface py-10"
  >
    <div
      class="container mx-auto px-4 md:px-6 lg:px-8"
    >
      <h2
        class="text-2xl md:text-3xl font-semibold text-center mb-10 text-white"
      >
        Compatible with Leading Frontend Frameworks
      </h2>
      <div
        class="flex flex-wrap justify-center items-center gap-8 mb-8"
      >
        <div
          class="w-1/3 md:w-auto"
        >
          <a
            class="block transition-all duration-300 hover:scale-105 hover:shadow-lg"
            data-testid="next-link"
            href="/frameworks/react"
            title="View details for React"
          >
            <div
              class="bg-neutral-light p-2 rounded-lg flex items-center justify-center h-20 md:h-24 md:w-48 border border-neutral-dark shadow-md"
            >
              <img
                alt="React logo"
                class="h-auto max-h-full w-auto"
                data-testid="next-image-React logo"
                height="70"
                src="/images/placeholders/react-logo.svg"
                width="120"
              />
            </div>
          </a>
        </div>
        <div
          class="w-1/3 md:w-auto"
        >
          <a
            class="block transition-all duration-300 hover:scale-105 hover:shadow-lg"
            data-testid="next-link"
            href="/frameworks/nextjs"
            title="View details for Next.js"
          >
            <div
              class="bg-neutral-light p-2 rounded-lg flex items-center justify-center h-20 md:h-24 md:w-48 border border-neutral-dark shadow-md"
            >
              <img
                alt="Next.js logo"
                class="h-auto max-h-full w-auto"
                data-testid="next-image-Next.js logo"
                height="80"
                src="/images/placeholders/nextjs-logo.png"
                width="120"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        class="flex flex-wrap justify-center items-center gap-6 md:gap-8 my-6"
      >
        <div
          class="w-1/4 md:w-auto"
        >
          <a
            class="block transition-all duration-300 hover:scale-105 hover:shadow-md"
            data-testid="next-link"
            href="/tools/tailwind"
            title="View details for Tailwind CSS"
          >
            <div
              class="bg-neutral-light p-3 rounded-lg flex items-center justify-center h-16 md:h-16 md:w-32 border border-neutral-dark shadow-sm"
            >
              <img
                alt="Tailwind CSS logo"
                class="h-auto max-h-full w-auto"
                data-testid="next-image-Tailwind CSS logo"
                height="40"
                src="/images/placeholders/tailwind-logo.svg"
                width="80"
              />
            </div>
          </a>
        </div>
        <div
          class="w-1/4 md:w-auto"
        >
          <a
            class="block transition-all duration-300 hover:scale-105 hover:shadow-md"
            data-testid="next-link"
            href="/tools/typescript"
            title="View details for TypeScript"
          >
            <div
              class="bg-neutral-light p-3 rounded-lg flex items-center justify-center h-16 md:h-16 md:w-32 border border-neutral-dark shadow-sm"
            >
              <img
                alt="TypeScript logo"
                class="h-auto max-h-full w-auto"
                data-testid="next-image-TypeScript logo"
                height="40"
                src="/images/placeholders/typescript-logo.svg"
                width="80"
              />
            </div>
          </a>
        </div>
        <div
          class="w-1/4 md:w-auto"
        >
          <a
            class="block transition-all duration-300 hover:scale-105 hover:shadow-md"
            data-testid="next-link"
            href="/frameworks/gatsby"
            title="View details for Gatsby"
          >
            <div
              class="bg-neutral-light p-3 rounded-lg flex items-center justify-center h-16 md:h-16 md:w-32 border border-neutral-dark shadow-sm"
            >
              <img
                alt="Gatsby logo"
                class="h-auto max-h-full w-auto"
                data-testid="next-image-Gatsby logo"
                height="40"
                src="/images/placeholders/gatsby-logo.svg"
                width="80"
              />
            </div>
          </a>
        </div>
        <div
          class="w-1/4 md:w-auto"
        >
          <a
            class="block transition-all duration-300 hover:scale-105 hover:shadow-md"
            data-testid="next-link"
            href="/frameworks/remix"
            title="View details for Remix"
          >
            <div
              class="bg-neutral-light p-3 rounded-lg flex items-center justify-center h-16 md:h-16 md:w-32 border border-neutral-dark shadow-sm"
            >
              <img
                alt="Remix logo"
                class="h-auto max-h-full w-auto"
                data-testid="next-image-Remix logo"
                height="40"
                src="/images/placeholders/remix-logo.svg"
                width="80"
              />
            </div>
          </a>
        </div>
        <div
          class="w-1/4 md:w-auto"
        >
          <a
            class="block transition-all duration-300 hover:scale-105 hover:shadow-md"
            data-testid="next-link"
            href="/frameworks/astro"
            title="View details for Astro"
          >
            <div
              class="bg-neutral-light p-3 rounded-lg flex items-center justify-center h-16 md:h-16 md:w-32 border border-neutral-dark shadow-sm"
            >
              <img
                alt="Astro logo"
                class="h-auto max-h-full w-auto"
                data-testid="next-image-Astro logo"
                height="40"
                src="/images/placeholders/astro-logo.svg"
                width="80"
              />
            </div>
          </a>
        </div>
      </div>
    </div>
  </section>
</div>
`;
